src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Pkcs12Builder.cs

```diff
@@ -0,0 +1,354 @@
+            if (pbeParameters.IterationCount < 1)
+                throw new ArgumentOutOfRangeException(nameof(pbeParameters.IterationCount));
+            if (safeContents.ConfidentialityMode != Pkcs12ConfidentialityMode.None)
+                throw new ArgumentException(SR.Cryptography_Pkcs12_CannotProcessEncryptedSafeContents, nameof(safeContents));
+            if (IsSealed)
+                throw new InvalidOperationException(SR.Cryptography_Pkcs12_PfxIsSealed);
+
+            PasswordBasedEncryption.ValidatePbeParameters(
+                pbeParameters,
+                password,
+                ReadOnlySpan<byte>.Empty);
+
+            byte[] encrypted = safeContents.Encrypt(password, ReadOnlySpan<byte>.Empty, pbeParameters);
+
+            if (_contents == null)
+            {
+                _contents = new List<ContentInfoAsn>();
+            }
+
+            _contents.Add(
+                new ContentInfoAsn
+                {
+                    ContentType = Oids.Pkcs7Encrypted,
+                    Content = encrypted,
+                });
+        }
+
+        public void AddSafeContentsUnencrypted(Pkcs12SafeContents safeContents)
+        {
+            if (safeContents == null)
+                throw new ArgumentNullException(nameof(safeContents));
+            if (IsSealed)
+                throw new InvalidOperationException(SR.Cryptography_Pkcs12_PfxIsSealed);
+
+            if (_contents == null)
+            {
+                _contents = new List<ContentInfoAsn>();
+            }
+
+            _contents.Add(safeContents.EncodeToContentInfo());
+        }
+
+        public byte[] Encode()
+        {
+            if (!IsSealed)
+            {
+                throw new InvalidOperationException(SR.Cryptography_Pkcs12_PfxMustBeSealed);
+            }
+
+            return _sealedData.ToArray();
+        }
+
+        public void SealWithMac(
+            string password,
+            HashAlgorithmName hashAlgorithm,
+            int iterationCount)
+        {
+            SealWithMac(
+                // This extension invoke allows null
+                password.AsSpan(),
+                hashAlgorithm,
+                iterationCount);
+        }
+
+        public void SealWithMac(
+            ReadOnlySpan<char> password,
+            HashAlgorithmName hashAlgorithm,
+            int iterationCount)
+        {
+            if (iterationCount < 1)
+                throw new ArgumentOutOfRangeException(nameof(iterationCount));
+            if (IsSealed)
+                throw new InvalidOperationException(SR.Cryptography_Pkcs12_PfxIsSealed);
+
+            ContentInfoAsn[] contents = _contents?.ToArray() ?? Array.Empty<ContentInfoAsn>();
+
+            byte[] rentedAuthSafe = null;
+            Span<byte> authSafeSpan = default;
+            byte[] rentedMac = null;
+            Span<byte> macSpan = default;
+            Span<byte> salt = stackalloc byte[0];
```

Comment:
I'm assuming the `= stackalloc byte[0]` is to avoid a compilation error about span lifetimes?  I'm curious in this case what the compiler complains about?

Reply:
System\Security\Cryptography\Pkcs\Pkcs12Builder.cs(191,28): error CS8353: A result of a stackalloc expression of type 'Span<byte>' cannot be used in this context because it may be exposed outside of the containing method

You can't assign a stackalloc in a different scope than the variable declaration unless the variable declaration included a stackalloc.

Reply:
> You can't assign a stackalloc in a different scope than the variable declaration unless the variable declaration included a stackalloc.

Yes, but that would only matter if it could possibly be returned, which doesn't appear to be the case here.  @VSadov, @jaredpar, is there a fundamental issue here, or the compiler just doesn't do the dataflow analysis to understand that this is actually safe?

Reply:
Based on the last time I talked to those guys about it: The compiler doesn't do the dataflow analysis.  A variable is either "returnable" or "not", and it's "returnable" unless it can very quickly prove otherwise.  @jaredpar had talked about some friendlier syntax here, like `Span<byte> salt = stackalloc;`, but that wouldn't help for the already released compilers.